# <center>Đồ án 3: Linear Regression</center>

# Thông tin sinh viên

- Họ và tên: Nguyễn Đình Văn
- MSSV: 20127662
- Lớp: 20CLC02

# Import

In [85]:
import pandas as pd
import numpy as np
import sklearn
# Import thêm dữ thư viện nếu cần

# Đọc dữ liệu

In [86]:
# Đọc dữ liệu bằng pandas
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)
X_train = train.iloc[:, :-1]    # Dataframe (chứa 10 đặc trưng huấn luyện)
y_train = train.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)

X_test = test.iloc[:, :-1]      # Dataframe (chứa 10 đặc trưng kiểm tra)
y_test = test.iloc[:, -1]       # Series    (chứa 1 giá trị mục tiêu kiểm tra)

listFreature = ['Adult Mortality', 'BMI', 'Polio', 'Diphtheria', 'HIV/AIDS', 'GDP', 'Thinness age 10-19', 'Thinness age 5-9', 'Income composition of resources', 'Schooling']
listModel = ['Sử dụng 2 đặc trưng (Adult Mortality, Schooling)', 'Sử dụng 3 đặc trưng (Adult Mortality, BMI, Schooling)', 'Sử dụng 2 đặc trưng (Adult Mortality, Schooling^2)']
# Sinh viên có thể sử dụng các khác nếu cần

# Cài đặt hàm

In [87]:
# Cài đặt các hàm cần thiết ở đây
class OLSLinearRegression:
    def fit(self, X, y):
        X_pinv = np.linalg.inv(X.T @ X) @ X.T    # np.linalg.pinv(X)
        self.w = X_pinv @ y

        return self

    def get_params(self):
        return self.w

    def predict(self, X):     
        return np.sum(self.w.ravel() * X, axis=1)   

def rmse(y, y_hat):
    return np.sqrt(np.mean((y.ravel() - y_hat.ravel())**2))

def find_best_feature ():
    # Xáo trộn 
    train_ = sklearn.utils.shuffle(train)
    train_ = train_.reset_index(drop = True)
    X_train_clone = train_.iloc[:, :-1]    # Dataframe (chứa 10 đặc trưng huấn luyện)
    y_train_clone = train_.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)

    RMSE_feature = list()
    for i in range(10):
        X_train_feature = X_train_clone.iloc[:,i].copy()
        y_train_feature = y_train_clone.copy()
        RMSE_ = 0
        for j in range(5):
            start = j * ((len(X_train_feature)) // 5) 
            end = (j + 1) * ((len(X_train_feature)) // 5)
            X_train_feature_ = X_train_feature.drop(X_train_feature.index[start:end])
            X_train_feature_ = X_train_feature_.reset_index(drop = True)

            Y_train_feature_ = y_train_feature.drop(y_train_feature.index[start:end])
            Y_train_feature_ = Y_train_feature_.reset_index(drop = True)

            X_val = X_train_feature[start:end].values.reshape(-1, 1)
            Y_val = y_train_feature[start:end].copy()

            lm = OLSLinearRegression()
            lm.fit(X_train_feature_.values.reshape(-1, 1), Y_train_feature_)
            y_pred = lm.predict(X_val)
            RMSE_ += rmse(Y_val, y_pred)

        RMSE_feature.append(RMSE_ / 5)
    return RMSE_feature

def find_best_model ():
    #Xáo trộn 
    train_ = sklearn.utils.shuffle(train)
    train_ = train_.reset_index(drop = True)
    X_train_clone = train_.iloc[:, :-1]    # Dataframe (chứa 10 đặc trưng huấn luyện)
    y_train_clone = train_.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)
    
    RMSE_model = list()
    X_train_1 = X_train_clone.iloc[:, [0,9]].copy()
    y_train_1 = y_train_clone.copy()

    X_train_2 = X_train_clone.iloc[:, [0,1,9]].copy()
    y_train_2 = y_train_clone.copy()

    X_train_3 = X_train_clone.iloc[:, [0,9]].copy()
    X_train_3.iloc[:,1] = X_train_3.iloc[:, 1] * X_train_3.iloc[:, 1]
    y_train_3 = y_train_clone.copy()

    #print(X_train_3, y_train_3)
    RMSE_1 = 0
    RMSE_2 = 0
    RMSE_3 = 0
    for i in range(5):
        start = i * ((len(X_train_1)) // 5) 
        end = (i + 1) * ((len(X_train_1)) // 5)

        X_train_1_ = X_train_1.drop(X_train_1.index[start:end])
        X_train_1_ = X_train_1_.reset_index(drop = True)
        Y_train_1_ = y_train_1.drop(y_train_1.index[start:end])
        Y_train_1_ = Y_train_1_.reset_index(drop = True)
        X_val_1 = X_train_1[start:end]
        Y_val_1 = y_train_1[start:end].copy()

        X_train_2_ = X_train_2.drop(X_train_2.index[start:end])
        X_train_2_ = X_train_2_.reset_index(drop = True)
        Y_train_2_ = y_train_2.drop(y_train_2.index[start:end])
        Y_train_2_ = Y_train_2_.reset_index(drop = True)
        X_val_2 = X_train_2[start:end]
        Y_val_2 = y_train_2[start:end].copy()

        X_train_3_ = X_train_3.drop(X_train_3.index[start:end])
        X_train_3_ = X_train_3_.reset_index(drop = True)
        Y_train_3_ = y_train_3.drop(y_train_3.index[start:end])
        Y_train_3_ = Y_train_3_.reset_index(drop = True)
        X_val_3 = X_train_3[start:end]
        Y_val_3 = y_train_3[start:end].copy()

        lm_1 = OLSLinearRegression()
        lm_1.fit(X_train_1_, Y_train_1_)
        y_pred_1 = lm_1.predict(X_val_1)
        RMSE_1 += rmse(Y_val_1, y_pred_1)
        
        lm_2 = OLSLinearRegression()
        lm_2.fit(X_train_2_, Y_train_2_)
        y_pred_2 = lm_2.predict(X_val_2)
        RMSE_2 += rmse(Y_val_2, y_pred_2)
        
        lm_3 = OLSLinearRegression()
        lm_3.fit(X_train_3_, Y_train_3_)
        y_pred_3 = lm_3.predict(X_val_3)
        RMSE_3 += rmse(Y_val_3, y_pred_3)
        
    RMSE_model.append(RMSE_1 / 5)
    RMSE_model.append(RMSE_2 / 5)
    RMSE_model.append(RMSE_3 / 5)
    return RMSE_model


# Yêu cầu 1a: Sử dụng toàn bộ 10 đặc trưng đề bài cung cấp (2 điểm) 

In [88]:
# Phần code cho yêu cầu 
# 1a
lm = OLSLinearRegression()
lm.fit(X_train, y_train)    
y_pred = lm.predict(X_test) 
hs = lm.get_params()            

In [89]:
# Gọi hàm RMSE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra
rmse(y_test, y_pred)

7.06404643058411

In [90]:
print('Life expectancy = ')
for i in range(len(hs)):
    if i != len(hs) - 1:
        print(listFreature[i],' * ', hs[i], ' + ')
    else:
        print(listFreature[i],' * ', hs[i])

Life expectancy = 
Adult Mortality  *  0.015101362735293323  + 
BMI  *  0.09021998065777065  + 
Polio  *  0.04292181752550481  + 
Diphtheria  *  0.13928911689487952  + 
HIV/AIDS  *  -0.567332827088243  + 
GDP  *  -0.00010076511487489812  + 
Thinness age 10-19  *  0.740713437758729  + 
Thinness age 5-9  *  0.1909357976739552  + 
Income composition of resources  *  24.505973591152646  + 
Schooling  *  2.3935166078326415


Công thức hồi quy

$$\text{Life expectancy} = {Adult Mortality} * 0.015101 + {BMI} * 0.090220 + {Polio} * 0.042922 + {Diphtheria} * 0.139289 - {HIV/AIDS} * 0.567333$$
$$\text - {GDP} *
 0.000101 + {Thinness age 10-19} * 0.740713 + {Thinness age 5-9} * 0.190936 + {Income composition of resources} * 24.505974 $$
$$\text - {GDP} + {Schooling} * 2.393517$$

# Yêu cầu 1b: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng, tìm mô hình cho kết quả tốt nhất (2 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [91]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu
RMSE_Feature = find_best_feature()
Pos = 0
for i in range(len(RMSE_Feature)):
    if (min(RMSE_Feature) == RMSE_Feature[i]):
        Pos = i

data = list(zip([name for name in listFreature], RMSE_Feature))
df = pd.DataFrame(data, columns=['Đặc trưng', 'RMSE'])
format_df = df.style.set_properties(**{'text-align': 'center'})
format_df =format_df.set_table_styles([dict(selector = 'th', props=[('text-align', 'center')])])
display(format_df)

In [92]:
# Huấn luyện lại mô hình best_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện
lm = OLSLinearRegression()
lm.fit(X_train.iloc[:, Pos].values.reshape(-1, 1), y_train)
y_pred = lm.predict(X_test.iloc[:, Pos].values.reshape(-1, 1))
hs = lm.get_params()    

In [93]:
# Gọi hàm RMSE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_feature_model
rmse(y_test, y_pred)

10.260950391655376

In [94]:
print('Life expectancy = ', listFreature[Pos],' * ', hs[-1])

Life expectancy =  Schooling  *  5.557399397691922


Công thức hồi quy

$$\text{Life expectancy} = {Schooling} * 5.5573994$$

# Yêu cầu 1c: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

In [95]:
# Phần code cho yêu cầu 1c
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu
RMSE_Model = find_best_model()
Pos = 0
for i in range(len(RMSE_Model)):
    if (min(RMSE_Model) == RMSE_Model[i]):
        Pos = i

data = list(zip([name for name in listModel], RMSE_Model))
df = pd.DataFrame(data, columns=['Mô hình', 'RMSE'])
format_df = df.style.set_properties(**{'text-align': 'center'})
format_df =format_df.set_table_styles([dict(selector = 'th', props=[('text-align', 'center')])])
display(format_df)

In [96]:
# Huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện

if Pos == 0: # y = Adult Mortality * x1 + Schooling *x2
    X_train_model = X_train.iloc[:, [0,9]].copy()
    y_train_model = y_train.copy()
    X_test_model = X_test.iloc[:, [0,9]].copy()
if Pos == 1: # y = Adult Mortality * x1 + BMI * x2 + Schooling *x3
    X_train_model = X_train.iloc[:, [0,1,9]].copy()
    y_train_model = y_train.copy()
    X_test_model = X_test.iloc[:, [0,1,9]].copy()
if Pos == 2: # y = Adult Mortality * x1 + Schooling^2 *x2
    X_train_model = X_train.iloc[:, [0,9]].copy()
    X_train_model.iloc[:,1] = X_train_model.iloc[:, 1] * X_train_model.iloc[:, 1]
    y_train_model = y_train.copy()
    X_test_model = X_test.iloc[:, [0,9]].copy()
    X_test_model.iloc[:,1] = X_test_model.iloc[:, 1] * X_test_model.iloc[:, 1]

lm = OLSLinearRegression()
lm.fit(X_train_model, y_train_model)    # Huaans luyen
y_pred = lm.predict(X_test_model) # 
hs = lm.get_params()             # lay ra he so 
        

In [97]:
# Gọi hàm RMSE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model
rmse(y_test, y_pred)

9.312027694213468

In [98]:
if Pos == 0:
    print('Life expectancy = ', listFreature[0],' * ', hs[0], ' + ', listFreature[9],' * ', hs[1])
if Pos == 1:
    print('Life expectancy = ', listFreature[0],' * ', hs[0], ' + ', listFreature[1],' * ', hs[1], ' + ', listFreature[9],' * ', hs[2])
if Pos == 2:
    print('Life expectancy = ', listFreature[0],' * ', hs[0], ' + ', listFreature[9],'^2 * ', hs[1])

Life expectancy =  Adult Mortality  *  0.02801217213290835  +  BMI  *  0.06604903698389963  +  Schooling  *  5.004187374704645


Công thức hồi quy

$$\text{Life expectancy} = {Adult Mortality} * 0.028012 + {BMI} * 0.066049 + {Schooling} * 5.004187$$